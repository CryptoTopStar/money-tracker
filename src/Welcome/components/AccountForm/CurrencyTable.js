import React from 'react'
import CurrencyInput from '../CurrencyInput'
import { CURRENCY } from '../../constants'
import { Checkbox, Table } from 'semantic-ui-react'

const CurrencyTable = ({ currencies, balance, onChecked, updateBalance }) => (
  <Table basic="very" celled>
    <Table.Header>
      <Table.Row>
        <Table.HeaderCell style={{ width: '50%' }}>
          Currency
        </Table.HeaderCell>
        <Table.HeaderCell>Initial Balance</Table.HeaderCell>
      </Table.Row>
    </Table.Header>
    <Table.Body>
      {currencies.map(code => {
        return (
          <Table.Row key={code}>
            <Table.Cell>
              <Checkbox
                checked={balance[code] !== undefined}
                onChange={onChecked}
                value={code}
                label={CURRENCY[code].name}
              />
            </Table.Cell>
            <Table.Cell>
              <CurrencyInput
                value={balance[code]}
                code={code}
                update={updateBalance}
              />
            </Table.Cell>
          </Table.Row>
        )
      })}
    </Table.Body>
  </Table>
)

export default CurrencyTable
