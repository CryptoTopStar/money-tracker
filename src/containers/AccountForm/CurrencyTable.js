import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Checkbox, Table } from 'semantic-ui-react'
import CurrencyInput from '../../components/CurrencyInput'
import { CURRENCY } from '../../constants/currency'
import {
  changeCurrencyCheckbox,
  changeCurrencyBalance
} from '../../actions/ui/accountForm'

class CurrencyTable extends React.Component {
  componentWillMount() {
    this.updateCurrencyList(this.props.base, this.props.secondary)
    this.setInitialBalanceCheckbox(this.props.base, this.props.balance)
  }

  componentWillReceiveProps(nextProps) {
    this.updateCurrencyList(nextProps.base, nextProps.secondary)
    this.setInitialBalanceCheckbox(nextProps.base, nextProps.balance)
  }

  updateCurrencyList(base, secondary) {
    this.currencies = [base, ...secondary]
  }

  setInitialBalanceCheckbox(base, balance) {
    if (Object.keys(balance).length === 0) {
      this.props.changeCurrencyCheckbox(base, true)
    }
  }

  handleCheckbox = (event, { value, checked }) => {
    this.props.changeCurrencyCheckbox(value, checked)
  }

  handleInput = code => event => {
    this.props.changeCurrencyBalance(code, event.target.rawValue)
  }

  render() {
    return (
      <Table basic="very" celled>
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell style={{ width: '50%' }}>
              Currency
            </Table.HeaderCell>
            <Table.HeaderCell>Initial Balance</Table.HeaderCell>
          </Table.Row>
        </Table.Header>
        <Table.Body>
          {this.currencies.map(code => {
            return (
              <Table.Row key={code}>
                <Table.Cell>
                  <Checkbox
                    checked={this.props.balance[code] !== undefined}
                    onChange={this.handleCheckbox}
                    value={code}
                    label={CURRENCY[code].name}
                  />
                </Table.Cell>
                <Table.Cell>
                  <CurrencyInput
                    value={this.props.balance[code]}
                    code={code}
                    onChange={this.handleInput(code)}
                  />
                </Table.Cell>
              </Table.Row>
            )
          })}
        </Table.Body>
      </Table>
    )
  }
}

CurrencyTable.propTypes = {
  base: PropTypes.string,
  secondary: PropTypes.arrayOf(PropTypes.string),
  balance: PropTypes.objectOf(PropTypes.string),
  changeCurrencyCheckbox: PropTypes.func,
  changeCurrencyBalance: PropTypes.func
}

const mapStateToProps = state => ({
  base: state.settings.currency.base,
  secondary: state.settings.currency.secondary,
  balance: state.ui.accountForm.balance
})

export default connect(mapStateToProps, {
  changeCurrencyCheckbox,
  changeCurrencyBalance
})(CurrencyTable)
