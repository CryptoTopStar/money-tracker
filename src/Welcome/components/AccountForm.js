import React, { Component } from 'react'
import { Form } from 'semantic-ui-react'
import CurrencyTable from './AccountForm/CurrencyTable'
import { accountTypesAsDropdownOptions } from '../constants/account'

class AccountForm extends Component {
  constructor(props) {
    super(props)
    this.types = accountTypesAsDropdownOptions()
    this.initialState = {
      accountName: '',
      accountType: this.types[0].key,
      balance: {
        [props.currencies[0]]: ''
      }
    }
    this.state = this.initialState
  }

  handleChange = (e, { name, value }) => this.setState({ [name]: value })

  onUsedCurrencyChange = (e, { checked, value }) => {
    this.setState(prevState => {
      const prevBalance = prevState.balance
      if (!checked) {
        // Prevent switching off the last active checkbox
        let activeBoxes = Object.keys(prevBalance).reduce((active, code) => {
          return prevBalance[code] !== undefined ? ++active : active
        }, 0)
        if (activeBoxes === 1) return
      }

      return {
        balance: { ...prevBalance, [value]: checked ? '' : undefined }
      }
    })
  }

  updateBalance = (code, value) => {
    this.setState(prevState => ({
      balance: { ...prevState.balance, [code]: value }
    }))
  }

  onAccountFormSubmit = e => {
    e.preventDefault()
    this.props.createAccount(this.state)
    this.setState(this.initialState)
  }

  render() {
    return (
      <Form onSubmit={this.onAccountFormSubmit}>
        <Form.Group widths="equal">
          <Form.Input
            label="Name"
            required
            name="accountName"
            placeholder="Account name"
            value={this.state.accountName}
            onChange={this.handleChange}
          />
          <Form.Select
            label="Type"
            name="accountType"
            placeholder="Account type"
            options={this.types}
            value={this.state.accountType}
            onChange={this.handleChange}
          />
        </Form.Group>
        <CurrencyTable
          currencies={this.props.currencies}
          balance={this.state.balance}
          onChecked={this.onUsedCurrencyChange}
          updateBalance={this.updateBalance}
        />
        <Form.Button content="Create Account" floated="right" />
      </Form>
    )
  }
}

export default AccountForm
