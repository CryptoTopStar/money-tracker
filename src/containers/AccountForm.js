import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Form } from 'semantic-ui-react'
import CurrencyTable from './AccountForm/CurrencyTable'
import { accountGroupAsDropdownOptions } from '../constants/account'
import { changeName, changeGroup } from '../actions/ui/accountForm'
import { createAccount } from '../actions/accounts'

class AccountForm extends React.Component {
  constructor(props) {
    super(props)

    this.groups = accountGroupAsDropdownOptions()
  }

  handleNameChange = (event, { value }) => this.props.changeName(value)
  handleGroupChange = (event, { value }) => this.props.changeGroup(value)

  handleSubmit = event => {
    event.preventDefault()
    this.props.createAccount(
      this.props.name,
      this.props.group,
      this.props.balance
    )
  }

  render() {
    return (
      <Form onSubmit={this.handleSubmit}>
        <Form.Group widths="equal">
          <Form.Input
            label="Name"
            required
            placeholder="Account name"
            value={this.props.name}
            onChange={this.handleNameChange}
          />
          <Form.Select
            label="Group"
            options={this.groups}
            value={this.props.group}
            onChange={this.handleGroupChange}
          />
        </Form.Group>
        <CurrencyTable />
        <Form.Button content="Create Account" floated="right" />
      </Form>
    )
  }
}

AccountForm.propTypes = {
  name: PropTypes.string,
  group: PropTypes.string,
  balance: PropTypes.objectOf(PropTypes.string),
  changeName: PropTypes.func,
  changeGroup: PropTypes.func,
  createAccount: PropTypes.func
}

const mapStateToProps = state => ({
  name: state.ui.accountForm.name,
  group: state.ui.accountForm.group,
  balance: state.ui.accountForm.balance
})

export default connect(mapStateToProps, {
  changeName,
  changeGroup,
  createAccount
})(AccountForm)
