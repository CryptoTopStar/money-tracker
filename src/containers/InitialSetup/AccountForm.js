import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Form } from 'semantic-ui-react'
import CurrencyTable from './CurrencyTable'
import Account from '../../models/Account'
import { changeName, changeGroup } from '../../actions/ui/accountForm'
import { saveAccount } from '../../actions/accounts'

class AccountForm extends React.Component {
  constructor(props) {
    super(props)

    this.groups = Account.groupOptions()
  }

  handleNameChange = (event, { value }) => this.props.changeName(value)
  handleGroupChange = (event, { value }) => this.props.changeGroup(value)

  handleSubmit = event => {
    event.preventDefault()
    this.props.saveAccount(Account.fromForm(this.props.form))
  }

  render() {
    return (
      <Form onSubmit={this.handleSubmit} style={{ marginBottom: '1em' }}>
        <Form.Group widths="equal">
          <Form.Input
            label="Name"
            required
            placeholder="Account name"
            value={this.props.form.name}
            onChange={this.handleNameChange}
          />
          <Form.Select
            label="Group"
            options={this.groups}
            value={this.props.form.group}
            onChange={this.handleGroupChange}
          />
        </Form.Group>
        <CurrencyTable />
        <Form.Button content="Create Account" />
      </Form>
    )
  }
}

AccountForm.propTypes = {
  form: PropTypes.shape({
    name: PropTypes.string,
    group: PropTypes.string,
    balance: PropTypes.objectOf(PropTypes.string)
  }),
  changeName: PropTypes.func,
  changeGroup: PropTypes.func,
  saveAccount: PropTypes.func
}

const mapStateToProps = state => ({
  form: state.ui.accountForm
})

export default connect(mapStateToProps, {
  changeName,
  changeGroup,
  saveAccount
})(AccountForm)
