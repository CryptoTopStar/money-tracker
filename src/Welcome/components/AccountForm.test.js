import React from 'react'
import renderer from 'react-test-renderer'
import AccountForm from './AccountForm'
global.navigator = {
  userAgent: 'node.js'
}

it('should render account form', () => {
  const tree = renderer
    .create(
      <AccountForm currencies={['USD', 'JPY']} createAccount={() => {}} />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
