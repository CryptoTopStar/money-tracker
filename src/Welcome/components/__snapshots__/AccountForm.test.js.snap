// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render account form 1`] = `
<form
  className="ui form"
  onSubmit={[Function]}
>
  <div
    className="equal width fields"
  >
    <div
      className="required field"
    >
      <label
        htmlFor={undefined}
      >
        Name
      </label>
      <div
        className="ui input"
      >
        <input
          name="accountName"
          onChange={[Function]}
          placeholder="Account name"
          required={true}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        htmlFor={undefined}
      >
        Type
      </label>
      <div
        aria-busy={undefined}
        aria-disabled={undefined}
        aria-expanded={false}
        aria-multiselectable={undefined}
        className="ui selection dropdown"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        required={undefined}
        role="listbox"
        tabIndex={0}
        type={undefined}
      >
        <select
          aria-hidden="true"
          multiple={undefined}
          name="accountType"
          type="hidden"
          value="cash"
        >
          <option
            value=""
          />
          <option
            value="cash"
          >
            Cash
          </option>
          <option
            value="bank"
          >
            Bank
          </option>
          <option
            value="deposit"
          >
            Deposit
          </option>
          <option
            value="credit"
          >
            Credit
          </option>
          <option
            value="asset"
          >
            Asset
          </option>
        </select>
        <div
          className="text"
        >
          Cash
        </div>
        <i
          aria-hidden="true"
          className="dropdown icon"
        />
        <div
          className="menu transition"
        >
          <div
            aria-checked={true}
            aria-disabled={undefined}
            aria-selected={true}
            className="active selected item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Cash
            </span>
          </div>
          <div
            aria-checked={false}
            aria-disabled={undefined}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Bank
            </span>
          </div>
          <div
            aria-checked={false}
            aria-disabled={undefined}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Deposit
            </span>
          </div>
          <div
            aria-checked={false}
            aria-disabled={undefined}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Credit
            </span>
          </div>
          <div
            aria-checked={false}
            aria-disabled={undefined}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Asset
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <table
    className="ui celled very basic table"
  >
    <thead
      className=""
    >
      <tr
        className=""
      >
        <th
          className=""
          style={
            Object {
              "width": "50%",
            }
          }
        >
          Currency
        </th>
        <th
          className=""
        >
          Initial Balance
        </th>
      </tr>
    </thead>
    <tbody
      className=""
    >
      <tr
        className=""
      >
        <td
          className=""
        >
          <div
            className="ui checked checkbox"
            onChange={[Function]}
            onClick={[Function]}
            onMouseDown={[Function]}
          >
            <input
              checked={true}
              className="hidden"
              name={undefined}
              readOnly={true}
              tabIndex={0}
              type="checkbox"
              value="USD"
            />
            <label>
              US Dollar
            </label>
          </div>
        </td>
        <td
          className=""
        >
          <div
            className="ui fluid left labeled input"
          >
            <div
              className="ui label mono"
              onClick={[Function]}
            >
              USD
            </div>
            <input
              data-cleave-ignore={
                Array [
                  "",
                  Object {
                    "initValue": "",
                    "numeral": true,
                    "numeralDecimalScale": 2,
                  },
                  undefined,
                  [Function],
                  undefined,
                  undefined,
                ]
              }
              name="USD"
              onChange={[Function]}
              onKeyDown={[Function]}
              type="text"
              value=""
            />
          </div>
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className=""
        >
          <div
            className="ui checkbox"
            onChange={[Function]}
            onClick={[Function]}
            onMouseDown={[Function]}
          >
            <input
              checked={false}
              className="hidden"
              name={undefined}
              readOnly={true}
              tabIndex={0}
              type="checkbox"
              value="JPY"
            />
            <label>
              Japanese yen
            </label>
          </div>
        </td>
        <td
          className=""
        >
          <div
            className="ui fluid left labeled input"
          >
            <div
              className="ui label mono"
              onClick={[Function]}
            >
              JPY
            </div>
            <input
              data-cleave-ignore={
                Array [
                  "",
                  Object {
                    "initValue": "",
                    "numeral": true,
                    "numeralDecimalScale": 0,
                  },
                  undefined,
                  [Function],
                  undefined,
                  undefined,
                ]
              }
              name="JPY"
              onChange={[Function]}
              onKeyDown={[Function]}
              type="text"
              value=""
            />
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="field"
  >
    <button
      className="ui button right floated"
      onClick={[Function]}
      required={undefined}
      tabIndex={undefined}
      type={undefined}
    >
      Create Account
    </button>
  </div>
</form>
`;
